In [1]:
# Update sklearn to prevent version mismatches
#!pip install sklearn --upgrade

In [2]:
# install joblib. This will be used to save your model. 
# Restart your kernel after installing 
#!pip install joblib

In [2]:
import pandas as pd
import numpy as np

Import our white and red wine data from the csvs

# Read the CSV and Perform Basic Data Cleaning

In [3]:
red_wine_df = pd.read_csv("winequality-red.csv", sep=';')
white_wine_df = pd.read_csv("winequality-white.csv", sep=';')
white_wine_df.info()
red_wine_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
fixed acidity           4898 non-null float64
volatile acidity        4898 non-null float64
citric acid             4898 non-null float64
residual sugar          4898 non-null float64
chlorides               4898 non-null float64
free sulfur dioxide     4898 non-null float64
total sulfur dioxide    4898 non-null float64
density                 4898 non-null float64
pH                      4898 non-null float64
sulphates               4898 non-null float64
alcohol                 4898 non-null float64
quality                 4898 non-null int64
dtypes: float64(11), int64(1)
memory usage: 459.3 KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
fixed acidity           1599 non-null float64
volatile acidity        1599 non-null float64
citric acid             1599 non-null float64
residual sugar          1599 non-null float64

## Unbalanced (Denormalized) Data for White Wine Data (4898 data points) and Red Wine Data (1599 data points) 

We want to use all the white and red wine data baseline condition we can use:

In [6]:
#if we want our only features of the model to consist of all white and red wine data, we use this data frame:
baseline_red_df = red_wine_df.copy()
baseline_white_df = white_wine_df.copy()
baseline_red_df["color"] = "red"
baseline_white_df["color"] = "white"
baseline_wine_df = pd.concat([baseline_red_df, baseline_white_df])
color_wine_df = pd.get_dummies(baseline_wine_df,columns=['color'])
colors_df = color_wine_df.drop(columns=["color_red"]).copy()
dummy_colors_df = colors_df.rename(columns={"color_white":"color"})
dummy_colors_df.to_csv("unbalanced-red-white-wine")

In [15]:
dummy_colors_df.groupby('quality').count()

fixed acidity  volatile acidity  citric acid  residual sugar  \
quality                                                                 
3                   30                30           30              30   
4                  216               216          216             216   
5                 2138              2138         2138            2138   
6                 2836              2836         2836            2836   
7                 1079              1079         1079            1079   
8                  193               193          193             193   
9                    5                 5            5               5   

         chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
quality                                                                        
3               30                   30                    30       30    30   
4              216                  216                   216      216   216   
5             2138                 2138                  2138     2138  2138   
6             2836                 2836                  2836     2836  2836   
7             1079                 1079                  1079     1079  1079   
8              193                  193                   193      193   193   
9                5                    5                     5        5     5   

         sulphates  alcohol  color  
quality                             
3               30       30     30  
4              216      216    216  
5             2138     2138   2138  
6             2836     2836   2836  
7             1079     1079   1079  
8              193      193    193  
9                5        5      5

In [29]:
#Suppose we want to take specific ranges in only_balanced_df, which does not include colors
range_colors_df = dummy_colors_df.loc[\
(dummy_colors_df["quality"] <= 8 ) & \
(dummy_colors_df["quality"] >= 3 )  \
] 


In [72]:
range_colors_df.groupby('quality').count()

fixed acidity  volatile acidity  citric acid  residual sugar  \
quality                                                                 
3                   30                30           30              30   
4                  216               216          216             216   
5                 2138              2138         2138            2138   
6                 2836              2836         2836            2836   
7                 1079              1079         1079            1079   
8                  193               193          193             193   

         chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
quality                                                                        
3               30                   30                    30       30    30   
4              216                  216                   216      216   216   
5             2138                 2138                  2138     2138  2138   
6             2836                 2836                  2836     2836  2836   
7             1079                 1079                  1079     1079  1079   
8              193                  193                   193      193   193   

         sulphates  alcohol  color  
quality                             
3               30       30     30  
4              216      216    216  
5             2138     2138   2138  
6             2836     2836   2836  
7             1079     1079   1079  
8              193      193    193

In [60]:
no_colors_df = range_colors_df.drop(columns="color")
no_colors_df

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4              0.70         0.00             1.9      0.076   
1               7.8              0.88         0.00             2.6      0.098   
2               7.8              0.76         0.04             2.3      0.092   
3              11.2              0.28         0.56             1.9      0.075   
4               7.4              0.70         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
3         9.8        6  
4         9.4        5  
...       ...      ...  
4893     11.2        6  
4894      9.6        5  
4895      9.4        6  
4896     12.8        7  
4897     11.8        6  

[6492 rows x 12 columns]

In [106]:
no_colors_df

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4              0.70         0.00             1.9      0.076   
1               7.8              0.88         0.00             2.6      0.098   
2               7.8              0.76         0.04             2.3      0.092   
3              11.2              0.28         0.56             1.9      0.075   
4               7.4              0.70         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
3         9.8        6  
4         9.4        5  
...       ...      ...  
4893     11.2        6  
4894      9.6        5  
4895      9.4        6  
4896     12.8        7  
4897     11.8        6  

[6492 rows x 12 columns]

In [61]:
no_colors_df.groupby("quality").count()

fixed acidity  volatile acidity  citric acid  residual sugar  \
quality                                                                 
3                   30                30           30              30   
4                  216               216          216             216   
5                 2138              2138         2138            2138   
6                 2836              2836         2836            2836   
7                 1079              1079         1079            1079   
8                  193               193          193             193   

         chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
quality                                                                        
3               30                   30                    30       30    30   
4              216                  216                   216      216   216   
5             2138                 2138                  2138     2138  2138   
6             2836                 2836                  2836     2836  2836   
7             1079                 1079                  1079     1079  1079   
8              193                  193                   193      193   193   

         sulphates  alcohol  
quality                      
3               30       30  
4              216      216  
5             2138     2138  
6             2836     2836  
7             1079     1079  
8              193      193

Add buckets to given range of quality values

### Moved quality values into buckets

Created bins for balanced data called: "terrible (0), mediocre (1), and great (2)"
The quality scores range from 0 to 4, 5 to 6, and 7 to 10. We will call them 0,1,2.

In [108]:
bin_balanced_df = no_colors_df.copy()
bins = [0, 4, 6, 8]
group_names = ["terrible", "mediocre", "great"]
bin_balanced_df["bin_quality"] = pd.cut(bin_balanced_df["quality"], bins, labels=group_names)
bin_qual_df = bin_balanced_df.drop(columns="quality")
dummy_bins_df = pd.get_dummies(bin_qual_df, drop_first=True)
dummy_bins_df
# columns=['bin_quality']
# pd.get_dummies(df, drop_first=True)
# bin_balanced_df.sample(30)


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4              0.70         0.00             1.9      0.076   
1               7.8              0.88         0.00             2.6      0.098   
2               7.8              0.76         0.04             2.3      0.092   
3              11.2              0.28         0.56             1.9      0.075   
4               7.4              0.70         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  bin_quality_mediocre  bin_quality_great  
0         9.4                     1                  0  
1         9.8                     1                  0  
2         9.8                     1                  0  
3         9.8                     1                  0  
4         9.4                     1                  0  
...       ...                   ...                ...  
4893     11.2                     1                  0  
4894      9.6                     1                  0  
4895      9.4                     1                  0  
4896     12.8                     0                  1  
4897     11.8                     1                  0  

[6492 rows x 13 columns]

In [47]:
bin_qual_df.groupby("bin_quality").count()

fixed acidity  volatile acidity  citric acid  residual sugar  \
bin_quality                                                                 
terrible               114               114          114             114   
mediocre              2509              2509         2509            2509   
great                  575               575          575             575   

             chlorides  free sulfur dioxide  total sulfur dioxide  density  \
bin_quality                                                                  
terrible           114                  114                   114      114   
mediocre          2509                 2509                  2509     2509   
great              575                  575                   575      575   

               pH  sulphates  alcohol  
bin_quality                            
terrible      114        114      114  
mediocre     2509       2509     2509  
great         575        575      575

Now we want to add colors with the convert quality scores to bins to see what happens

In [18]:
bins_colors_df = dummy_rename_colors_df.copy()
bins = [0, 4, 6, 10]
group_names = [0, 1, 2]
bins_colors_df["bin_quality"] = pd.cut(bins_colors_df["quality"], bins, labels=group_names)
new_bins_colors_df = bins_colors_df.drop(columns="quality")
new_bins_colors_df

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4              0.70         0.00             1.9      0.076   
1               7.8              0.88         0.00             2.6      0.098   
2               7.8              0.76         0.04             2.3      0.092   
3              11.2              0.28         0.56             1.9      0.075   
4               7.4              0.70         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.7              0.34         0.31            16.4      0.051   
1595            9.3              0.31         0.49             1.3      0.042   
1596            6.4              0.17         0.27             6.7      0.036   
1597            7.5              0.29         0.36            15.7      0.050   
1598            6.8              0.28         0.29            11.9      0.052   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 20.0                 146.0  0.99834  3.06       0.54   
1595                 34.0                 147.0  0.99480  3.11       0.46   
1596                 88.0                 223.0  0.99480  3.28       0.35   
1597                 29.0                 124.0  0.99680  3.06       0.54   
1598                 51.0                 149.0  0.99544  3.02       0.58   

      alcohol  colors bin_quality  
0         9.4       0           1  
1         9.8       0           1  
2         9.8       0           1  
3         9.8       0           1  
4         9.4       0           1  
...       ...     ...         ...  
1594      9.1       1           1  
1595      9.8       1           1  
1596     10.2       1           1  
1597     10.4       1           1  
1598     10.4       1           1  

[3198 rows x 13 columns]

In [109]:
dummy_bins_df

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4              0.70         0.00             1.9      0.076   
1               7.8              0.88         0.00             2.6      0.098   
2               7.8              0.76         0.04             2.3      0.092   
3              11.2              0.28         0.56             1.9      0.075   
4               7.4              0.70         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  bin_quality_mediocre  bin_quality_great  
0         9.4                     1                  0  
1         9.8                     1                  0  
2         9.8                     1                  0  
3         9.8                     1                  0  
4         9.4                     1                  0  
...       ...                   ...                ...  
4893     11.2                     1                  0  
4894      9.6                     1                  0  
4895      9.4                     1                  0  
4896     12.8                     0                  1  
4897     11.8                     1                  0  

[6492 rows x 13 columns]

# Select your features (columns)

In [128]:
# Set features. Drop quality, y value. 
X = dummy_bins_df.drop(columns=['bin_quality_great'])
# .drop(columns=['bin_quality_mediocre', 'bin_quality_great'])
X.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  bin_quality_mediocre  
0      9.4                     1  
1      9.8                     1  
2      9.8                     1  
3      9.8                     1  
4      9.4                     1

# Create a Train Test Split

Use `quality` for the y values

In [129]:
from sklearn.model_selection import train_test_split
#80% train, 20% test. y is this one column
y = dummy_bins_df[['bin_quality_great']]
#random state 42 will have same picks for x test and y test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [130]:
X_train.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
2932            6.4              0.25         0.32            11.3      0.038   
1539            7.2              0.39         0.32             1.8      0.065   
1109           10.8              0.47         0.43             2.1      0.171   
100             8.3              0.61         0.30             2.1      0.084   
2816            7.2              0.17         0.41             1.6      0.052   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
2932                 69.0                 192.0  0.99573  3.14       0.50   
1539                 34.0                  60.0  0.99714  3.46       0.78   
1109                 27.0                  66.0  0.99820  3.17       0.76   
100                  11.0                  50.0  0.99720  3.40       0.61   
2816                 24.0                 126.0  0.99228  3.19       0.49   

      alcohol  bin_quality_mediocre  
2932     10.2                     1  
1539      9.9                     1  
1109     10.8                     1  
100      10.2                     1  
2816     10.8                     1

# Train the Model using Random Forest 



In [131]:
#Train the model
from sklearn import tree
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier()
rfc.fit(X_train, y_train)

/opt/anaconda3/envs/PythonData/lib/python3.7/site-packages/ipykernel_launcher.py:5: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  """


RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [132]:
#Test the model
from sklearn.metrics import confusion_matrix, classification_report
predictions = rfc.predict(X_test)
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       1.00      0.99      0.99      1309
           1       0.95      0.98      0.97       314

    accuracy                           0.99      1623
   macro avg       0.97      0.99      0.98      1623
weighted avg       0.99      0.99      0.99      1623



# Save the Model

In [36]:
# save your model by updating "your_name" with your name
# and "your_model" with your model variable
# be sure to turn this in to BCS
# if joblib fails to import, try running the command to install in terminal/git-bash
import joblib
filename = 'unbalanced-with-color.sav'
joblib.dump(predictions, filename)

['unbalanced-with-color.sav']